macro/src/lib.rs

```diff
@@ -23,6 +23,26 @@ use subxt_codegen::{
 };
 use syn::{parse_macro_input, punctuated::Punctuated};
 
+use std::path::Path;
+
+use subwasmlib::{source::Source, OutputFormat, Subwasm};
+
+fn process_wasm_file(wasm_path: &Path) -> Result<Vec<u8>, CodegenError> {
+    fn to_error<A: ToString>(err: A) -> CodegenError {
+        CodegenError::Fetch(subxt_codegen::error::FetchMetadataError::Other(
+            err.to_string(),
+        ))
+    }
+    let source =
+        Source::from_options(Some(wasm_path.to_path_buf()), None, None, None).map_err(to_error)?;
+    let subwasm = Subwasm::new(&source.try_into().map_err(to_error)?).map_err(to_error)?;
+    let mut output = vec![];
+    subwasm
+        .write_metadata(OutputFormat::Scale, None, &mut output)
```

Comment:
Ideally, it would be interesting if we could:
- decode the metadata in a way that we don't depend on substrate crates (`sp-*` `sc_executor`)
- extract from the wasm blob just the `Metadata_metadata()` bytes

If that turns out to be too complicated, we can still look into obtaining that information via substrate-crates. Using subwasmlib to fetch the metadata brings quite a tree of dependencies.

For extracting the wasm blob, we'd need to rely on `sp_maybe_compressed_blob`:
 https://github.com/chevdor/subwasm/blob/fae3e13bd1f8680bf32cbf2ed223223583f8d713/libs/wasm-loader/src/lib.rs#L152

For calling into the wasm blob and fetching the metadata, we'd rely on`sc-executor`:
https://github.com/chevdor/subwasm/blob/fae3e13bd1f8680bf32cbf2ed223223583f8d713/libs/wasm-testbed/src/lib.rs#L131-L147

Reply:
rewritten to use direct dependencies in last commit